cvmfs/cache_ram.cc

```diff
@@ -141,59 +206,156 @@ void RamCacheManager::CtrlTxn(
+      transaction->buffer = realloc(transaction->buffer, transaction->size);
+      if (!transaction->buffer) {
+        LogCvmfs(kLogCache, kLogDebug, "realloc failed: %s", strerror(errno));
+        return -errno;
+      }
+    } else {
+      LogCvmfs(kLogCache, kLogDebug,
+               "attempted to write more than requested (%u>%u)",
+               size, transaction->size);
+      return -ENOSPC;
+    }
+  }
+
+
+  uint64_t copy_size = min(size, transaction->size - transaction->pos);
+  LogCvmfs(kLogCache, kLogDebug, "copy %u bytes of transaction %s",
+           copy_size, transaction->id.ToString().c_str());
+  memcpy(static_cast<char *>(transaction->buffer) + transaction->pos,
+         buf, copy_size);
+  transaction->pos += copy_size;
+  perf::Inc(counters_.n_write);
+  return copy_size;
 }
 
 
 int RamCacheManager::Reset(void *txn) {
   Transaction *transaction = reinterpret_cast<Transaction *>(txn);
   transaction->pos = 0;
+  LogCvmfs(kLogCache, kLogDebug, "reset transaction %s",
+           transaction->id.ToString().c_str());
+  perf::Inc(counters_.n_reset);
   return 0;
 }
 
 
 int RamCacheManager::OpenFromTxn(void *txn) {
-  Transaction *transaction = reinterpret_cast<Transaction *>(txn);
-  int64_t handle = CommitToKvStore(transaction);
-  if (handle < 0)
-    return handle;
-  ReadOnlyFd file_descriptor(handle, transaction->pos);
-
   WriteLockGuard guard(rwlock_);
-  return AddFd(file_descriptor);
+  Transaction *transaction = reinterpret_cast<Transaction *>(txn);
+  int64_t retval = CommitToKvStore(transaction);
+  if (retval < 0) {
+    LogCvmfs(kLogCache, kLogDebug,
+             "error while commiting transaction on %s: %s",
+             transaction->id.ToString().c_str(), strerror(-retval));
+    return retval;
+  }
+  LogCvmfs(kLogCache, kLogDebug, "open pending transaction for %s",
+           transaction->id.ToString().c_str());
+  perf::Inc(counters_.n_openfromtxn);
+  return DoOpen(transaction->id);
 }
 
 
 int RamCacheManager::AbortTxn(void *txn) {
   Transaction *transaction = reinterpret_cast<Transaction *>(txn);
   if (transaction->buffer)
     free(transaction->buffer);
+  LogCvmfs(kLogCache, kLogDebug, "abort transaction %s",
+           transaction->id.ToString().c_str());
+  perf::Inc(counters_.n_aborttxn);
   return 0;
 }
 
 
 int RamCacheManager::CommitTxn(void *txn) {
+  WriteLockGuard guard(rwlock_);
   Transaction *transaction = reinterpret_cast<Transaction *>(txn);
-  int retval = CommitToKvStore(transaction);
-  return retval;
+  perf::Inc(counters_.n_committxn);
+  return CommitToKvStore(transaction);
 }
 
 
 int64_t RamCacheManager::CommitToKvStore(Transaction *transaction) {
-  if (transaction->handle >= 0)
-    return transaction->handle;
+  kvstore::MemoryBuffer buf;
+  buf.address = transaction->buffer;
+  kvstore::MemoryKvStore *store;
+  if (transaction->expected_size == kSizeUnknown) {
+    buf.size = transaction->pos;
+    buf.address = realloc(buf.address, buf.size);
```

Comment:
I think I'd only do the `realloc` and the check for `buf.address` if `buf.size > 0`.